### ISSUE TITLE
Anti-spam tools

### ISSUE THREAD

**Issue by fallanic** [echojs/echojs#13]
Ideas from this other thread https://github.com/echojs/echojs/issues/12

- email sign up
- captcha on sign up
- "Report Spam" option? (downvote might actually work better, but an admin view for downvoted posts might be helpful

**Comment by theabbie** [comment]
CAPTCHA should solve the issue as account creation won't be possible using a script, and, using VPN will be inefficient.

Maybe, peer moderation will be useful, if post recieves 3-4 downvotes, alert the Moderator.

**Comment by canton7** [comment]
Just a quick note that even recaptcha is pretty easy to bypass: there are entire farms of people dedicated to solving them on demand for a small fee (as I found out when I tried to use them to reduce spam...). API-accessible email addresses are 2 a penny IIRC.

You raise the bar a bit with measures like these, but I'd caution against expecting them to make any difference against someone with resources.

**Comment by LoganDark** [comment]
> I'd caution against expecting them to make any difference against someone with resources.

It's all about raising the bar. Nobody can reasonably expect to defeat someone with resources - if pretty much anyone poured in enough time, they could harass any small service as much as they wanted. ~~They could also look for people like theabbie posting working exploits in public without responsible disclosure~~

Story time!

I actually discovered an exploit a few years ago that allowed you to create mass amounts of Discord accounts, with no email address or password, and then use them like normal registered users. Pretty sure I was around 12 back then. (My account has since been deleted, so there's no more COPPA non-compliance.)

That's a nice example of how one person can cause disruption. If I had registered, say, 100 accounts, and then run bots on them with the then-far-too-capable discord.py library, I could have sent 1000 messages per second in any server that I chose. That was before Discord search, so any message caught in the crossfire would have been completely drowned out, and I would have stressed the crap out of that Discord server. The moderators would've had to ban 100 individual user accounts, making the whole situation a complete mess. I had far too much power and I was only 12. Needless to say, I (and my botnet) got deleted before I could use most of that power. That was _definitely_ for the better.

It was only possible because... Discord sometimes did not show a captcha when creating a new account, allowing me to use cheap free services like Runscope to make new accounts from all sorts of IP addresses that couldn't be traced back to me! _And_ Discord allowed you to create accounts _without_ filling in account information! On top of _that_, it allowed you to join the accounts to servers and send messages without ever actually authenticating over websocket! Absolute dream for any exploiter.

Yes, I did abuse the API significantly and I did not disclose anything to Discord. They had to track me down and ban me on their own. I regret that, but it's a cool story to tell, even though 12-year-old me was literally a toxic hacker. Every single API hole I've mentioned has since been patched, and they also added a captcha to the bot authorization screen after I collected every public bot on Discord in one server.

There are so many stars that aligned to allow that, and compromising any one of them would have hindered it significantly. The goal is to knock as many stars out of alignment as possible to make it as difficult as possible for someone with resources to assemble a proper attack strategy. That is the goal of captchas, new member limits and so on. And that would have prevented my 12 year old rampage.

Hopefully, that will also help this project prevent future 12 year old rampages from any other aspiring hackers. We'll just have to wait and see which measures are going to be put in place. :P

**Comment by canton7** [comment]
Agreed. I was just slightly surprised that after adding recaptcha to a form the spammers didn't even blink -- it seems whatever tooling they were using automatically detected the recaptcha and solved it.

It's a worthwhile thing to do, don't get me wrong, but it does seem faintly ironic that the only spammers it will realistically stop are kids who are spamming to try and make a point about there being no spam protection...

**Comment by theabbie** [comment]
@canton7 The effort put into spamming depends on the benefits from spamming, EchoJS isn't a very popular spamming choice, Solving RECAPTCHA definitely requires some software which isn't free, so, it rules out majority of spammers.

**Comment by LoganDark** [comment]
> Agreed. I was just slightly surprised that after adding recaptcha to a form the spammers didn't even blink -- it seems whatever tooling they were using automatically detected the recaptcha and solved it.
> 
> It's a worthwhile thing to do, don't get me wrong, but it does seem faintly ironic that the only spammers it will realistically stop are kids who are spamming to try and make a point about there being no spam protection...

Unfortunately, a recaptcha will not stop a real person who wants to spam, because they can solve captchas, even without renting time from a captcha farm.

It would be better to implement a trust based system where newer users are less privileged and can earn privileges by being a constructive and trustworthy member of the community. This could mean posting limits, pre-approval requirement (great for preventing drive-by spam by a bunch of new accounts), etc.

In my experience, despite all the data recaptcha sends to Google, recaptchas work almost entirely by IP address and anyone from a network that isn't flagged will be able to solve mass amounts of them with ease. I can't speak for other solutions like hCaptcha though. I know this because my home network is somehow flagged every time whenever I make the occasional Google search.

**Comment by LoganDark** [comment]
> @canton7 The effort put into spamming depends on the benefits from spamming, EchoJS isn't a very popular spamming choice, Solving RECAPTCHA definitely requires some software which isn't free, so, it rules out majority of spammers.

Usually people either spam because they like breaking things or spam because they like bothering the moderation / users. Maybe they have something to advertise. Hell, maybe they're selling their spambot.

You most likely fall under the first category. You abused something for fun, then you posted it publicly and probably continued to abuse it occasionally to "prove" your point that spammers exist. But I'm not accusing you of anything - like you say, it could have been anyone. Anyone who saw your post and got inspired.

**Comment by theabbie** [comment]
@LoganDark A RECAPTCHA triggers based on history of the IP address, mouse movements and stuff, so, a bot which can simulate human behaviour might bypass it. It doesn't prevent manual spam, but, there is IP address rate limiting as well, which means using VPN which still isn't free (and limited). Only a sophisticated spammer can successfully spam a CAPTCHA based Community. The goal is to make it difficult so that spammers don't bother to try.

**Comment by LoganDark** [comment]
> @LoganDark A RECAPTCHA triggers based on history of the IP address, mouse movements and stuff, so, a bot which can simulate human behaviour might bypass it. It doesn't prevent manual spam, but, there is IP address rate limiting as well, which means using VPN which still isn't free (and limited). Only a sophisticated spammer can successfully spam a CAPTCHA based Community. The goal is to make it difficult so that spammers don't bother to try.

Even on a CAPTCHA which demands 100% certainty (people are always required to pass challenges), any unsophisticated spammer with nothing but *time* can solve it repeatedly and spam a community, even if it's entirely implemented properly.

And any unsophisticated spammer can easily install one of those browser extensions which will automatically detect captchas and buy a solve from a captcha farm. Then they can be even _faster_ doing it manually! No sophistication needed.

I believe the only security measures that will work are measures that demand actual real contribution to the community, with lots of time and effort put in. It will take weeks or months to build up a reputation good enough to have the privilege of making a single spam post before you're demoted/banned.

You could also just make echojs uninteresting as an attack target, and less people will try to attack it.

**Comment by theabbie** [comment]
@LoganDark A spammer spending money to spam counts as sophisticated spammer, Click Farms are real people, so, it is impossible to prevent that, There are Communities with invite-only registration, that might be equivalent to reputation based system, and if a user spams, the person who invites him has to suffer, that would be effective.

**Comment by LoganDark** [comment]
> @LoganDark A spammer spending money to spam counts as sophisticated spammer, Click Farms are real people, so, it is impossible to prevent that,

It costs pennies on the dollar to solve a captcha - the usual price is $0.02 - $0.05 per captcha, according to my brief research a couple years ago. It's not very sophisticated - I've had an allowance of $10/week for a couple years now, some 12 year olds are lucky enough to have one too. Or they could use their mom's credit card.

> There are Communities with invite-only registration, that might be equivalent to reputation based system, and if a user spams, the person who invites him has to suffer, that would be effective.

That's actually a good idea. You could go the (iirc) VirusShare route and have each user have 25 invite slots. They can each invite up to 25 more people, and each of those people have their own 25 invite slots too. If there are mass amounts of abuse, there is a single point of failure for the entire botnet, or at least many single points of failure. You just keep nipping branches off the tree. One account banned, half the botnet is gone - another account banned, botnet is gone.

What's tricky is getting the moderation tools right for this - you need administrative control over the amount of invites someone has, in case you want to grant additional invites or something, you need to be able to walk the tree and find out who invited someone, who invited them, who someone has invited, etc. you need to be able to ban entire subtrees of accounts or apply limits/suspensions - echojs just doesn't sound like a place that kind of complexity should belong.

**Comment by theabbie** [comment]
@LoganDark [lobste.rs](https://lobste.rs) is a perfect example of such Community, I can say it's a 0 spam Community, and it has reputation system too, so, your actions are blocked based on your history. Spam doesn't persist long enough there. Implementing that on EchoJS will require complete redesign. Also, new users are given ability to invite more people after certain reputation. It can't be more perfect than that for legitimate users and can't be more worse than that for spammers.

**Comment by tracker1** [comment]
First, I have a [PR in to add docker-compose](https://github.com/echojs/echojs/pull/14) not being a Ruby developer, I spent a while Sunday getting this much working... from here, as I've mentioned in other posts, I'd like to re-write the site as-is in Node, if only so that I might have a more familiar environment, while keeping the initial db and api the same.

Email specification and validation for login would probably be the first thing to add after that, and would go a very long way.  As would blacklists for email or domain to prevent throw-aways or those with wildcards.  I don't think recaptcha would add value above email validation.  Could do a small script to ensure a modern browser with JS support (compute a hash on the client/browser that is somewhat expensive and can be verified against a server pool) based on email address.

After that, or with that would likely want to update the UI slightly to get it a little more mobile friendly.

**Comment by shadowtime2000** [comment]
I think that is the best way to go about this.

**Comment by theabbie** [comment]
@tracker1 If Email Authentication is added, will the existing accounts be forced to add email address or they can exist as-is?

**Comment by tracker1** [comment]
@theabbie afaik, that's the plan...

But instead of being patient and waiting for a free weekend or two for someone to work on this, you continue to throw tantrums and act like a petulant, spoiled child, while contributing nothing of value.

**Comment by theabbie** [comment]
@tracker1 What I am doing currently isn't harmful, it is just about sharing somewhat unrelated content. I won't attack the server anymore.

**Comment by shadowtime2000** [comment]
I don't think we could take your word for it... we have repeatedly told you to stop on EchoJS and in other places and you have not. @theabbie What you are doing is harmful because you are just kind of messing up the nice community that we have their, and if you didn't know, unrelated content is spam.

**Comment by theabbie** [comment]
@shadowtime2000 All this because I am doing this without being anonymous, EchoJS should be fine with unrelated content as it is moderated atleast once daily. Attacking server was bad, so, I am not doing that.

**Comment by shadowtime2000** [comment]
@theabbie EchoJS isn't fine with unrelated content because if you didn't know, the moderators actually have lives where they don't spend all day trying to spam on community driven sites. They have other things to do besides trying to fight against people like you who just try to mess it up. Just not posting unrelated content is better because we are then not putting as much work on the moderators.

**Comment by theabbie** [comment]
@shadowtime2000 If they are moderating once daily, then it doesn't make a difference, and moderating atleast once daily is a necessity, for any community.

**Comment by shadowtime2000** [comment]
@theabbie Yes moderating once daily is a great idea but that doesn't mean its fine to put unrelated content on it. Make their lives a little easier would you

**Comment by shadowtime2000** [comment]
> You most likely fall under the first category. You abused something for fun, then you posted it publicly and probably continued to abuse it occasionally to "prove" your point that spammers exist. But I'm not accusing you of anything - like you say, it could have been anyone. Anyone who saw your post and got inspired.

I would like point out he also kind of also falls under the second category of wanting to advertise something because he was desperate to get stars on his PixelCraft project.

**Comment by theabbie** [comment]
@shadowtime2000 I no longer want to prove any point, neither can I promote anything as it has become very obvious now. Even if anyone else spams this server, I will be blamed, and, of course, I deserve that. I should make Moderators life easier, so, no more EchoJS. Would love to see it when everything gets fixed.

**Comment by theabbie** [comment]
@shadowtime2000 Though I am not sure about it, I am somehow recieving Email subscription confirmation emails, and it's obvious that someone is trying to give me the taste of Spam, If it's you, stop. Email Spam is bad. And, If it's a lesson, I have already learned it.

**Comment by shadowtime2000** [comment]
> @shadowtime2000 Though I am not sure about it, I am somehow recieving Email subscription confirmation emails, and it's obvious that someone is trying to give me the taste of Spam, If it's you, stop. Email Spam is bad. And, If it's a lesson, I have already learned it.

@theabbie That most definitely is not me, as I have said before most people have better things to do than spam.

**Comment by tracker1** [comment]
It's not me... I'm not *that* petty or vindictive. As to Lobste.rs, it's invite based, and the github project has over 138 contributors... if people like @theabbie actually contributed code instead of destruction, it could be better already.

https://github.com/lobsters/lobsters

**Comment by tracker1** [comment]
@theabbie what you are doing, is posting off-topic content... You know you're being a childish a-hole, we all know it... just stop already.

**Comment by shadowtime2000** [comment]
I don't think an invite based system would work well, because believe it or not there aren't as many active users and it would be very hard for people who are just randos online (like me) to join.

**Comment by theabbie** [comment]
@shadowtime2000 Invite-based system was just a proposal, it's not possible to implement that here. Here, the goal is to reduce spam. Lobsters barely has any spam. Since the existing users are good enough, it would be better to disable registration till things get fixed. People can use EchoJS without accounts too. That would be reasonable as shutting it down doesn't sound correct.

**Comment by theabbie** [comment]
@tracker1 Fine, EchoJS has already suffered a lot, And, before comparing Lobsters to EchoJS, you should know that they are putting effort in that Community, There hasn't been any change in EchoJS since years, and there is a reason for that. EchoJS is close to abandoned and only exists on servers running by itself. Maybe it was a demo project by the creator "Salvatore Sanfilippo" to demonstrate Redis database. It has to change, maybe more people will be willing to contribute if anyone showed any effort.

**Comment by shadowtime2000** [comment]
I was right, @theabbie didn't stop spamming. I guess we can't trust people like @theabbie to keep their word.

@theabbie 🤦 there will probably be changes. You've shown us spam is a problem so now we are going to fix it. Also maybe now we will have a 6 person moderation team just to make sure you never join EchoJS again. Let's just protect the community from your childish attention wanting behavior.

**Comment by theabbie** [comment]
@theabbie Good For EchoJS, go for it.

**Comment by theabbie** [comment]
@shadowtime2000  Good For EchoJS, go for it.

**Comment by tracker1** [comment]
@theabbie so contribute in a positive way.. make the changes you want and submit pull requests... or are you just an a-hole script kiddie with nothing better to do?

**Comment by theabbie** [comment]
@tracker1 I will try to contribute, will take some time to understand this code.
